feat: JailedStream

ryanolson opened the pull request:
## Summary
This PR introduces `JailedStream`, a stream processing optimization for tool calling scenarios that improves performance and reliability while maintaining full OpenAI compatibility.

## Key Features
- **Lazy tool parsing**: Tool calls only parsed when jailing ends or stream completes
- **UTF-8 safe**: All string operations respect character boundaries
- **Multiple parser support**: Hermes, Mistral, Phi4, Llama3, and custom formats
- **Comprehensive edge case handling**: Partial matches, malformed JSON, early exits
- **Better streaming behavior**: Accumulates content for complete tool calls rather than partial processing

## Recent Major Improvements

### Async Removal Refactor
- **Root cause**: Harmony parser used `tokio::sync::OnceCell` instead of `std::sync::OnceLock`
- **Impact**: Simplified tool parsing chain by removing unnecessary async from several core functions
- **Result**: Cleaner and more maintainable code with synchronous tool parsing

### UTF-8 Safety Fixes
- Fixed character boundary violations in prefix matching
- Added comprehensive Unicode test coverage

## Usage
```rust
let jail = JailedStream::builder()
    .tool_call_parser("hermes")
    .build();

let optimized_stream = jail.apply(chat_completion_stream);
```

## Testing
- ✅ 26 comprehensive test cases covering all scenarios
- ✅ UTF-8 safety validated with Unicode edge cases
- ✅ All existing tests pass, clippy clean
- ✅ Full integration with existing OpenAI protocols

## Breaking Changes
None - this is a pure optimization that maintains full API compatibility.

elyasmnvidian committed:
feat: always stream for tool calling

Signed-off-by: ayushag <ayushag@nvidia.com>

elyasmnvidian committed:
chore: moved tool parsing to preprocessor.rs

Signed-off-by: ayushag <ayushag@nvidia.com>

elyasmnvidian committed:
chore: added unit tests

Signed-off-by: ayushag <ayushag@nvidia.com>

elyasmnvidian committed:
chore: rebase and updated function calls

Signed-off-by: ayushag <ayushag@nvidia.com>

ayushag-nv committed:
fix: curl doesn't break now

Signed-off-by: ayushag <ayushag@nvidia.com>

elyasmnvidian committed:
chore: enabled stream=true for tool choice

Signed-off-by: ayushag <ayushag@nvidia.com>

ayushag-nv committed:
chore: fix aggregator - clippy to be fixed

Signed-off-by: ayushag <ayushag@nvidia.com>

ayushag-nv committed:
fix: fixed rebased artifacts

Signed-off-by: ayushag <ayushag@nvidia.com>

biswapanda committed:
fix: dyn namespace scoping for trtllm (#2970)

Signed-off-by: Biswa Panda <biswa.panda@gmail.com>
Signed-off-by: ayushag <ayushag@nvidia.com>

ayushag-nv committed:
fix: fixed tests and rebase artifacts

Signed-off-by: ayushag <ayushag@nvidia.com>

ayushag-nv committed:
Merge branch 'main' into elyas/streamtool

ayushag-nv committed:
Merge branch 'main' into elyas/streamtool

ayushag-nv committed:
Merge branch 'main' into elyas/streamtool

ryanolson committed:
feat: add standalone JailedStream implementation for token jail detection

- Implement JailedStream with builder pattern for clean API
- Use async-stream's stream! macro for cleaner async implementation
- Support configurable jail start/end sequences
- Integrate tool call detection and parsing
- Add comprehensive tests with shared test utilities
- Create documentation for usage patterns

The JailedStream provides a standalone solution for accumulating tokens
when certain sequences are detected (jailing) and releasing them as a
single chunk when the jail ends, enabling proper tool call handling in
streaming responses.

Signed-off-by: Ryan Olson <rolson@nvidia.com>

ryanolson committed:
refactor: optimize JailedStream for better performance

- Remove unnecessary annotation code that was never used
- Fix efficiency issue by avoiding cloning when not jailing
- Only clone response data when actually entering jail state
- Simplify jailed behavior to not yield empty chunks
- Accumulate content silently and only yield final result
- Update tests to match new behavior

This makes the stream processing more efficient by:
1. Eliminating unnecessary clones during normal streaming
2. Removing pointless annotation objects
3. Not sending empty chunks while accumulating
4. Only yielding the final aggregated result when jail ends

Signed-off-by: Ryan Olson <rolson@nvidia.com>

ryanolson committed:
perf: optimize JailedStream to use impl Stream and remove context overhead

- Remove ManyOut type alias and AsyncEngineContextProvider dependency
- Change apply() to return impl Stream instead of Pin<Box<ResponseStream>>
- Remove unnecessary context extraction and re-wrapping
- Use tokio::pin!() for stack pinning instead of Box::pin() (more efficient)
- Simplify tests by removing mock context infrastructure
- Update documentation with correct usage examples

This makes JailedStream a pure stream combinator that:
1. Doesn't handle context (caller's responsibility)
2. Avoids unnecessary boxing in the library
3. Can be composed with other stream transformers
4. Lets callers decide when to add boxing/context

Performance improvements:
- No heap allocation for pinning (stack pinning instead)
- No context passing overhead
- Direct stream transformation without wrapper types

Signed-off-by: Ryan Olson <rolson@nvidia.com>

ryanolson committed:
refactor: update preprocessor to use new JailedStream implementation

- Remove old apply_tool_calling_jail_internal function
- Remove PossibleToolCallAnnotation struct and implementations
- Update apply_tool_calling_jail_with_parser to conditionally use JailedStream
- Only apply jail logic when tool_call_parser is configured
- Comment out old jail tests that are no longer applicable
- Clean up unused imports and run cargo fmt/clippy

ryanolson committed:
feat: add dual entry/exit paths for JailedStream

- Change jail entry logic from if/else to evaluate both sequence and tool call conditions
- Add early exit detection via should_exit_jail_early() method
- Support two paths to enter jail: explicit sequences OR tool call detection
- Support two paths to exit jail: end sequences OR complete tool call parsing
- Add tests for dual entry paths and early exit behavior
- Improve debug logging to show which condition triggered jail start/end

ryanolson committed:
refactor: optimize stream transformations to reduce boxing

- Separate context passing from stream transformations
- Change transform_embedding_postprocessor_stream to return impl Stream
- Update transform_postprocessor_stream to accept context as parameter
- Defer boxing until the final ResponseStream::new() call
- Extract context once at the beginning of generate() methods
- Reduce boxing from 5-6 locations to 1 per request
- Improves performance by eliminating intermediate heap allocations

ryanolson committed:
feat: add conditional tool jail application based on tool_choice

- Only apply jail when tools are present AND tool_choice is not None
- Error if tool_choice is 'required', 'auto', or named but no parser configured
- Skip jail entirely when tool_choice is explicitly None
- Add should_apply_tool_jail() to determine jail application logic
- Refactor to separate decision logic from stream processing
- Avoid lifetime issues by evaluating conditions before stream transformation

ryanolson committed:
chore: clean up commented-out tests and add gitignore changes

- Removed old jail implementation tests that are no longer compatible
- Kept test_detect_tool_call_start_different_parsers as it tests parser behavior
- Old tests were testing annotation behavior that no longer exists in new JailedStream
- New tests in jail.rs provide equivalent coverage

ryanolson committed:
Merge branch 'main' into ryan/streamtool

Resolved merge conflicts in .gitignore and preprocessor.rs:
- Fixed gitignore format for Ruler generated files
- Removed duplicate jail implementation from main
- Kept optimized JailedStream with conditional application
- Fixed transform_postprocessor_stream signature usage
- All tests passing

ryanolson committed:
fix: resolve clippy warnings and test compilation errors

- Remove unused imports and dead code warnings in test module
- Fix test_streaming_usage.rs function signature for transform_postprocessor_stream
- Remove test_preprocessor.rs that tested old jail implementation
- Add missing context parameter to transform calls
- All tests now pass and clippy/fmt are clean

ryanolson committed:
test: add comprehensive jail functionality test coverage

Add 10 new test cases for JailedStream covering all parser types and edge cases:
- Hermes parser with <tool_call> markers
- Mistral parser (both variants with and without [TOOL_CALLS] marker)
- Phi4 parser with <|tool_call|> markers
- llama3_json parser with <|python_tag|> markers
- False positive detection for non-tool JSON
- Malformed JSON handling
- Partial tool call scenarios
- Empty stream behavior
- Multiple consecutive tool calls
- Fragmented tool calls across many small chunks

All tests verify silent jailing behavior (no empty chunks) and proper
tool call accumulation between start/end markers.

ryanolson committed:
fix: preserve Annotated metadata through jail processing

Fixes issue where create_unjailed_response was discarding upstream correlation
metadata (id, event, comment) from Annotated wrapper, breaking observability.

Changes:
- Capture id/event/comment fields when jail is triggered
- Pass preserved metadata to create_unjailed_response
- Attach original metadata to synthesized responses
- Add comprehensive test coverage for metadata preservation

Tests added:
- test_jailed_stream_preserves_metadata: Normal jail processing
- test_jailed_stream_preserves_metadata_on_stream_end: Stream termination
- test_jailed_stream_metadata_edge_cases: Partial/empty metadata

Addresses GitHub issue discussion r2352198367

ryanolson commented on lib/llm/src/protocols/openai/chat_completions/jail.rs:
```diff
@@ -0,0 +1,612 @@
+            event: None,
+            comment: None,
+        }
+    }
```
Addressed in: e027e610

ryanolson committed:
fix: preserve trailing content after jail end markers

Fixes issue where trailing content in the same chunk as jail end markers
was being lost. When a chunk contains both an end marker and additional
content (e.g., "</tool_call>trailing text"), the trailing text was
discarded instead of being emitted as a separate chunk.

Changes:
- Find exact position of end markers for precise content splitting
- Split accumulated content at marker boundary into jailed/trailing parts
- Emit jailed content as tool call response, trailing as pass-through
- Handle both explicit end sequences and early exit scenarios
- Add parser-specific end position detection for all supported formats
- Preserve metadata in trailing content chunks

Tests added:
- test_jailed_stream_trailing_content_same_chunk: End marker + trailing
- test_jailed_stream_early_exit_with_trailing: Early exit + trailing

Fixes GitHub issue discussion r2352198360

ryanolson committed:
feat: refactor jail state to support independent multi-choice processing

Replace HashMap-based jail state with deterministic Vec-based collection
to enable independent choice jailing and ensure consistent processing order.

Key improvements:
- Add ChoiceJailState and ChoiceJailStateCollection for deterministic ordering
- Implement independent per-choice jail/unjail logic
- Add EmissionMode configuration (Packed vs SingleChoicePerChunk)
- Support configurable emission patterns for OpenAI compatibility
- Fix double emission issue during choice unjailing
- Maintain backward compatibility for single-choice scenarios

This addresses GitHub issues r2352198339 (determinism) and enables
proper multi-choice (n>1) tool call handling with independent state
management per choice index.

Tests: 22/24 pass (2 expected multi-choice test failures for future work)

ryanolson committed:
feat: implement independent multi-choice jailing architecture

This major refactoring enables true independent choice processing where each
choice (when n>1) can jail and unjail independently with deterministic ordering.

Core changes:
- Add ChoiceEmission enum for clean emission abstraction
- Add ChoiceJailState with encapsulated jail logic per choice
- Replace HashMap with deterministic Vec-based ChoiceJailStateCollection
- Implement independent process_content() method per choice
- Add configurable EmissionMode (Packed vs SingleChoicePerChunk)
- Clean separation between state management and emission strategy

Key achievements:
- Multi-choice independent jailing tests now pass
- Deterministic choice ordering (always 0, 1, 2...)
- Metadata preservation through jail processing
- Maintains backward compatibility for single-choice scenarios

Tests: 22/24 pass (2 trailing content edge cases remaining)

Addresses GitHub issues:
- r2352198339: HashMap determinism and independent choice jailing
- Enables proper multi-choice (n>1) tool call handling

ryanolson committed:
fix: separate trailing content emission for independent choice jailing

Modify emission grouping logic to separate trailing emissions from
tool/content emissions. This ensures trailing content after jail end
markers is always emitted as separate chunks, fixing the failing tests:
- test_jailed_stream_trailing_content_same_chunk
- test_jailed_stream_early_exit_with_trailing

Also remove unused methods to satisfy clippy warnings.

All 24 jail tests now pass with the independent multi-choice jailing
implementation.

Signed-off-by: Ryan Olson <rolson@nvidia.com>

ryanolson commented on lib/llm/src/protocols/openai/chat_completions/jail.rs:
```diff
@@ -0,0 +1,612 @@
+// SPDX-FileCopyrightText: Copyright (c) 2024-2025 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+use std::collections::HashMap;
```
this should be updated

ryanolson committed:
feat: implement partial marker matching for streaming tool calls

Add efficient prefix matching utility to detect tool call markers split
across chunk boundaries. Solves the issue where markers like "<TOOLCALL>"
were missed when characters arrived in separate chunks.

Key improvements:
- Created utils/prefix_matcher.rs with MarkerMatcher using Aho-Corasick
- Detects partial suffixes (e.g., "n<TO" finds "<TO" as partial match)
- Prevents false positives (e.g., "n < 5" doesn't trigger jailing)
- Preserves chunk boundaries to avoid unnecessary yield consumption
- Handles complete tool calls within single combined chunks
- Maintains backward compatibility with all existing functionality

Tests added:
- test_partial_matching_suffix_detection: Verifies cross-chunk detection
- test_partial_matching_false_positive_prevention: Prevents math expressions

All 26 existing jail tests continue to pass.

Signed-off-by: Ryan Olson <rolson@nvidia.com>

ryanolson committed:
refactor: standardize jail tests with human-readable assertions

Transform jail tests to use standardized Content()/ToolCall() notation
and comprehensive helper functions for better readability and debugging.

Key improvements:
- Added helper functions: assert_content, assert_tool_call, reconstruct_content
- Documented expected input→output transformation in each test
- Replaced complex .any()/.find() chains with index-based assertions
- Added content reconstruction verification to ensure no data loss
- Standardized test structure: chunk count → individual chunks → content reconstruction

Tests now serve as executable documentation with immediate visual clarity
of the stream processing behavior. When tests fail, developers know
exactly which chunk contains the unexpected result.

Updated tests:
- test_partial_matching_suffix_detection: Shows cross-chunk detection
- test_partial_matching_false_positive_prevention: Shows false positive handling
- test_jailed_stream_no_jailing: Shows pass-through behavior
- test_jailed_stream_empty_stream: Shows edge case handling
- test_jailed_stream_multiple_tool_calls: Shows complex multi-tool behavior
- test_jailed_stream_trailing_content_same_chunk: Shows trailing preservation

All 26 jail tests continue to pass with enhanced clarity and debuggability.

Signed-off-by: Ryan Olson <rolson@nvidia.com>

ryanolson commented:
@CoderRabbit full review

ryanolson committed:
refactor: standardize jail test assertions for improved readability

Replace complex .iter().any() chains with direct index-based assertions
in 11 jail stream tests. Updates include:

- Convert to assert_content() and assert_tool_call() helper functions
- Add explicit chunk count assertions (assert_eq!(results.len(), N))
- Document expected input→output transformations in test comments
- Use Content() and ToolCall() notation for clarity
- Add content reconstruction verification

Tests updated:
- test_jailed_stream_tool_call_across_many_chunks
- test_jailed_stream_early_exit
- test_jailed_stream_hermes_parser
- test_jailed_stream_mistral_parser
- test_jailed_stream_mistral_parser_with_tool_calls_marker
- test_jailed_stream_phi4_parser
- test_jailed_stream_llama3_json_parser
- test_jailed_stream_false_positive_json
- test_jailed_stream_malformed_tool_call
- test_jailed_stream_partial_tool_call
- test_jailed_stream_early_exit_with_trailing

Improves test maintainability and debugging - failures now show exact
chunk index and content differences rather than vague .any() mismatches.

Signed-off-by: Ryan Olson <rolson@nvidia.com>

ryanolson committed:
test: add comprehensive edge case tests for prefix matcher

Add 3 new test cases to verify correct behavior of multi-pattern
partial matching in streaming scenarios:

- test_multiple_partial_matches_edge_case: Verifies that invalid
  partial prefixes (like "FooBaz" for "FooBar") are correctly
  skipped in favor of valid ones

- test_earliest_valid_partial_match: Tests overlapping potential
  partials where only the later one is valid

- test_partial_at_exact_end: Ensures valid partials at string
  end are correctly detected

These tests confirm the find_partial_suffix algorithm correctly
handles complex cases where multiple patterns could potentially
match but only some are actually valid prefixes. The trie-based
validation ensures optimal partial detection for streaming
tool call marker recognition.

All tests pass, confirming the implementation is robust.

Signed-off-by: Ryan Olson <rolson@nvidia.com>

ryanolson commented:
/ok to test 3128bf2

ryanolson committed:
fix: implement UTF-8 safe slicing in prefix matcher

- Add safe_slice helper to ensure slicing only at character boundaries
- Fix complete match handling to use UTF-8 safe slicing instead of byte indices
- Fix partial match handling to use UTF-8 safe slicing
- Update find_partial_suffix to iterate over char boundaries instead of bytes
- Fix return type ambiguity in should_apply_tool_jail function
- Add comprehensive unicode tests to prevent regressions
- Resolves panics when patterns occur at multi-byte character boundaries

Fixes code review issues from PR #3034

Signed-off-by: Ryan Olson <rolson@nvidia.com>

ryanolson commented on lib/llm/src/utils/prefix_matcher.rs:
```diff
@@ -0,0 +1,432 @@
+            }
+        }
+        None
+    }
```
Addressed in : 5158aea

ryanolson commented on lib/llm/src/utils/prefix_matcher.rs:
```diff
@@ -0,0 +1,432 @@
+                marker_start: mat.start(),
+                suffix: combined[mat.end()..].to_string(),
+            };
+        }
```
Addressed in : 5158aea

ryanolson commented:
/ok to test 5158aea

ryanolson committed:
refactor: remove unnecessary async from tool parsing chain

The tool parsing functions were unnecessarily async due to using
tokio::sync::OnceCell instead of std::sync::OnceLock for Harmony
encoding initialization. This cascaded async requirements through
the entire tool parsing and jail processing chain.

Changes:
- Replace tokio::sync::OnceCell with std::sync::OnceLock in harmony parser
- Remove async from harmony parser functions
- Remove async from try_tool_call_parse and detect_and_parse_tool_call
- Remove async from try_tool_call_parse_aggregate/stream functions
- Remove async from jail processing functions (create_tool_call_choice, etc.)
- Remove unnecessary .await calls throughout the chain
- Fix type inference issue in jail tests

The Harmony encoding is now loaded synchronously on first use, which is
appropriate since it's a one-time initialization that only blocks the
first caller. This significantly simplifies the codebase by removing
100+ unnecessary async/await points.

All tests passing, no functional changes to tool parsing behavior.

Signed-off-by: Ryan Olson <rolson@nvidia.com>

ayushag-nv commented on lib/llm/src/protocols/openai/chat_completions/jail.rs:
```diff
@@ -0,0 +1,2647 @@
+
+    /// Find the exact position where the tool call ends for splitting content
+    /// This handles the early exit case where we have trailing content after the tool call
+    fn find_tool_call_end_position(&self, content: &str, parser: &str) -> usize {
```
@ryanolson I think parser library should be the correct place for any parser specific bits. Can you move it there ?

ayushag-nv commented on lib/llm/src/utils/prefix_matcher.rs:
```diff
@@ -0,0 +1,562 @@
+    }
+
+    #[test]
+    fn test_partial_buffer_combination() {
```
This one is super amazing

ayushag-nv commented on lib/llm/src/preprocessor.rs:
```diff
@@ -1090,3 +869,183 @@ impl
 }
+
+#[allow(deprecated, dead_code)]
+#[cfg(test)]
```
ok can we move this to jail also ?

GuanLuo reviewed (commented):
Submitted partial comments. Will review `prefix_matcher.rs` next

ayushag-nv commented on lib/llm/src/protocols/openai/chat_completions/jail.rs:
```diff
@@ -0,0 +1,2647 @@
+                    marker,
+                    suffix,
+                    ..
+                } => {
```
@GuanLuo haha , we are kind of leaving similar comments for Ryan

ryanolson commented on lib/llm/src/protocols/openai/chat_completions/jail.rs:
```diff
@@ -0,0 +1,2647 @@
+            // Track Annotated metadata for preservation
+            let mut last_annotated_id: Option<String> = None;
+            let mut last_annotated_event: Option<String> = None;
+            let mut last_annotated_comment: Option<Vec<String>> = None;
```
It's essentially annotated.rs in runtime.

It matches up to how one might use the SSE response stream if one wanted to be completely disjoint from oai's usage.

The proper way to use it is like OAI's responses API.  @paulhendricks is a good person to chat with on that.

ayushag-nv commented:
@ryanolson There is one method in `detect_tool_call_start` in parser library, that was detecting the tool calling start earlier. Now that has been superseded by prefix_matcher. Do you mind cleaning those pieces up as well as thats basically some more dead code.

ryanolson commented on lib/llm/src/utils/prefix_matcher.rs:
```diff
@@ -0,0 +1,562 @@
+    }
+
+    #[test]
+    fn test_earliest_valid_partial_match() {
+        // Test that the algorithm finds the earliest VALID partial match
+        // Patterns: ["FooBar", "<TOOLCALL>"]
+        // Input: "Some text FooBa and then <TO"
+        // Analysis: "FooBa and then <TO" is not a valid prefix of "FooBar" because
+        // after "FooBa" we have " " (space) but "FooBar" expects "r"
+        // Expected: Skip invalid "FooBa..." and find valid "<TO" partial
+        let patterns = vec!["FooBar".to_string(), "<TOOLCALL>".to_string()];
+        let matcher = MarkerMatcher::new(patterns).unwrap();
+
+        let result = matcher.process_chunk("Some text FooBa and then <TO", "");
+
+        if let MatchResult::Partial {
+            prefix,
+            partial,
+            possible_patterns,
+        } = result
+        {
+            // Should find "<TO" as the valid partial match
+            assert_eq!(partial, "<TO");
+            assert_eq!(prefix, "Some text FooBa and then ");
+            assert!(possible_patterns.contains(&"<TOOLCALL>".to_string()));
+        } else {
+            panic!("Expected partial match for '<TO>', got: {:?}", result);
+        }
+    }
```
that was the idea.  i wanted to see what happens if a partial match was started, but then failed to complete and a different pattern was matched later.

ryanolson commented:
> @ryanolson There is one method in `detect_tool_call_start` in parser library, that was detecting the tool calling start earlier. Now that has been superseded by prefix_matcher. Do you mind cleaning those pieces up as well as thats basically some more dead code.

Honestly, I tried not to look at what that method did, as I was only concerned about what happened when the start was triggered.

If prefix_matcher supercedes it, I still think we can leave it it.  It could offer us a programmatic path for more complex pattern matches like regex or other patterns where we might provide it with a closure.

ayushag-nv commented on lib/llm/src/preprocessor.rs:
```diff
@@ -662,262 +620,53 @@ impl OpenAIPreprocessor {
+    ) -> std::result::Result<bool, Error> {
+        match (tool_call_parser, tool_choice, has_tools) {
+            // No parser but tools requested - error cases
+            (None, Some(ChatCompletionToolChoiceOption::Required), true) => Err(anyhow::anyhow!(
```
This is a UX decision. @GuanLuo You have any thoughts on this -- how it should behave ??

GuanLuo replied:
This is really up to interpretation. As documented in tool choice option: `required` means the model must call one or more tools. I think **logging warning** is fine in the case where the model does return text containing tool calling message, not having a tool parser only affect that the information is not transformed to tool call field.

elyasmnvidian reviewed (commented):
@ryanolson 

In the coming week we want to add a feature to detect raw JSON tool calling where there isn't any tool call start token or end token. This happens in some mistral models and llama 3.1 sometimes.

We need some way to detect "{":
1.  emit it in content if it doesn't lead to valid JSON OR if parsing fails in the tool parser on the JSON
2. emit it in tools response if parsing succeeds and don't emit in content field

I am not sure if this PR handles the case where if the tool parser fails to parse the JSON (even if its valid JSON) we decide to emit it in content or not. We should write extensive test cases for these.

Also I think one of the files has test cases that has to be moved to another file / its own file.

elyasmnvidian commented:
We also plan on doing parallel tool calling.

Parallel tool calls may or may not have a "<PARALLEL TOOL START>" or "<PARALLEL TOOL END>"

Parallel tool calls are required to have a tool call start token, end token may be optional.

<PARALLEL TOOL START>
<TOOL_CALL>{....}</TOOL_CALL>
<TOOL_CALL>{....}</TOOL_CALL>
<TOOL_CALL>{....}</TOOL_CALL>
<PARALLEL TOOL END>

This can come in a future MR and will tackle after this is merged.

GuanLuo commented on lib/llm/src/utils/prefix_matcher.rs:
```diff
@@ -0,0 +1,562 @@
+//! This module provides utilities for detecting complete and partial marker patterns
+//! in streaming text, with support for detecting markers split across chunk boundaries.
+
+use aho_corasick::{AhoCorasick, AhoCorasickBuilder, MatchKind};
```
Moving to parser library makes sense

GuanLuo commented:
> We also plan on doing parallel tool calling.
> 
> ```
> Parallel tool calls may or may not have a "<PARALLEL TOOL START>" or "<PARALLEL TOOL END>"
> 
> Parallel tool calls are required to have a tool call start token, end token may be optional.
> 
> <PARALLEL TOOL START>
> <TOOL_CALL>{....}</TOOL_CALL>
> <TOOL_CALL>{....}</TOOL_CALL>
> <TOOL_CALL>{....}</TOOL_CALL>
> <PARALLEL TOOL END>
> 
> This can come in a future MR and will tackle after this is merged.
> ```

Wouldn't this already be handled? As long as the jail_start/end_sequence contains `<PARALLEL TOOL START>` and `<PARALLEL TOOL END>`. Actually I guess the tricky part is that now `</TOOL_CALL>` should not unjail the stream if the stream is jailed by `<PARALLEL TOOL START>`

GuanLuo commented:
Let's create a future item to also reuse the JailStream concept on reasoning parser as well. Feel really trivial if we replace "tool call" stuff with a trait

ryanolson committed:
fix: address PR review comments for JailedStream

- Remove redundant _enable_tool_calling variable assignment
- Remove obsolete maybe_enable_tool_call function superseded by should_apply_tool_jail
- Add documentation for common tool call markers including harmony parser
- Auto-determine jail sequences from parser config when not manually set
- Fix async/await issues in parser tests after async removal
- Add #[allow(dead_code)] for unused test utility functions

Addresses PR review comments:
- 2360290856: Remove redundant code at line 699-700
- 2360342550: Remove obsolete function superseded by new logic
- 2360590684: Document common markers behavior with specific parsers
- 2360770209: Add harmony parser documentation
- 2360813802: Auto-populate jail sequences from parser configuration
- 2360839391: Fix test handling for missing sequences (now auto-populated)

Signed-off-by: Ryan Olson <rolson@nvidia.com>

ryanolson commented:
Addressed in 081126b7:

🚀 Rocket emoji comments addressed:
- 2360290856 (Line 699): ✅ Removed redundant `_enable_tool_calling` variable assignment
- 2360342550 (Line 742): ✅ Removed obsolete `maybe_enable_tool_call` function (lines 68-80)
- 2360590684 (Line 882): ✅ Added documentation clarifying common markers are always included
- 2360770209 (Line 891): ✅ Added harmony parser documentation in common markers list
- 2360813802 (Line 913): ✅ Auto-populate jail sequences from parser config when not manually set
- 2360839391 (Line 668): ✅ Test now works correctly with auto-populated sequences from parser

All rocket-marked review comments have been addressed in this commit.

ryanolson commented:
> Let's create a enhancement ticket to also reuse the JailStream concept on reasoning parser as well. Feel really trivial if we replace "tool call" stuff with a trait

Same bits work for stop words. 

I think we should move this all to the Backend.

ayushag-nv commented on lib/llm/src/protocols/openai/chat_completions/jail.rs:
```diff
@@ -0,0 +1,2667 @@
+            // Add JSON start patterns for Mistral-style tool calls
+            "[{".to_string(),
+            "{".to_string(),
+            // Note: Harmony parser uses JSON patterns, covered by "{" above
```
@ryanolson Harmony has the wierd tags in their format like channel nd stuff. I am not sure how { covers that up

ayushag-nv commented on lib/llm/src/protocols/openai/chat_completions/jail.rs:
```diff
@@ -0,0 +1,2667 @@
+            let parser_map = get_tool_parser_map();
+            if let Some(config) = parser_map.get(parser_name.as_str()) {
+                // Auto-populate start sequences if none configured
+                if self.jail_start_sequences.is_empty() {
+                    self.jail_start_sequences = config.json.tool_call_start_tokens.clone();
+                }
+
+                // Auto-populate end sequences if none configured
+                if self.jail_end_sequences.is_empty() {
+                    self.jail_end_sequences = config.json.tool_call_end_tokens.clone();
+                }
+            }
```
@ryanolson Can you extend the tests also for different models without setting jail start sequences manually. Wanted to check if whatever we had in configs properly covers this logic or not.

ayushag-nv committed:
chore: move tests to test_jail.rs

Signed-off-by: ayushag <ayushag@nvidia.com>

elyasmnvidian committed:
chore: update cargo lock

Signed-off-by: Elyas Mehtabuddin <emehtabuddin@nvidia.com>

elyasmnvidian committed:
Merge branch 'main' into ryan/streamtool

ayushag-nv committed:
fix: bugs

Signed-off-by: ayushag <ayushag@nvidia.com>

elyasmnvidian committed:
chore: fix unit test #1

Signed-off-by: Elyas Mehtabuddin <emehtabuddin@nvidia.com>

ayushag-nv committed:
fix: more bugs

Signed-off-by: ayushag <ayushag@nvidia.com>

ayushag-nv committed:
Merge branch 'main' into ryan/streamtool

ayushag-nv committed:
fix: clippy

Signed-off-by: ayushag <ayushag@nvidia.com>

elyasmnvidian committed:
chore: fix await unit test for harmony

Signed-off-by: Elyas Mehtabuddin <emehtabuddin@nvidia.com>

ayushag-nv committed:
fix: harmony

Signed-off-by: ayushag <ayushag@nvidia.com>

ayushag-nv committed:
Merge branch 'main' into ryan/streamtool

elyasmnvidian committed:
preprocessor: warn and proceed when no parser configured for tool_choice; jail: gate common markers when parser is set and clarify local naming; tests: tighten dual-entry path assertions

ayushag-nv committed:
Merge branch 'main' into ryan/streamtool

ayushag-nv commented:
/ok test f58c24ab64be9819313177fdc6a703f9ec291f72

ayushag-nv committed:
fix: cargo fmt

Signed-off-by: ayushag <ayushag@nvidia.com>

ayushag-nv committed:
fix: fmt parsers

Signed-off-by: ayushag <ayushag@nvidia.com>

ayushag-nv commented:
/ok to test e04a1cd

ayushag-nv committed:
fix: ci bugs

Signed-off-by: ayushag <ayushag@nvidia.com>

ayushag-nv committed:
Merge branch 'main' into ryan/streamtool

elyasmnvidian committed:
chore: fix await unit test for harmony #2

Signed-off-by: Elyas Mehtabuddin <emehtabuddin@nvidia.com>

elyasmnvidian committed:
fix: revert back to Grahams fix and make jail async

Signed-off-by: Elyas Mehtabuddin <emehtabuddin@nvidia.com>

elyasmnvidian committed:
chore: clippy #3

Signed-off-by: Elyas Mehtabuddin <emehtabuddin@nvidia.com>
